Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+                                return;
+                            }
+                        }
+                        result = visitor.EndArray(arrayContainer.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, NodePtr>)
+                {
+                    const Node& node = *AZStd::get<NodePtr>(m_value);
+                    result = visitor.StartNode(node.GetName());
+                    if (result.IsSuccess())
+                    {
+                        const Object::ContainerType& object = GetObjectInternal();
+                        for (const Object::EntryType& entry : object)
+                        {
+                            result = visitor.Key(entry.first);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                            result = entry.second.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+
+                        const Array::ContainerType& arrayContainer = GetArrayInternal();
+                        for (const Value& entry : arrayContainer)
+                        {
+                            result = entry.Accept(visitor, copyStrings);
+                            if (!result.IsSuccess())
+                            {
+                                return;
+                            }
+                        }
+
+                        result = visitor.EndNode(object.size(), arrayContainer.size());
+                    }
+                }
+                else if constexpr (AZStd::is_same_v<Alternative, AZStd::any>)
+                {
+                    result = visitor.OpaqueValue(*arg);
+                }
+            },
+            m_value);
+        return result;
+    }
+
+    AZStd::unique_ptr<Visitor> Value::GetWriteHandler()
+    {
+        return AZStd::make_unique<ValueWriter>(*this);
+    }
+
+    bool Value::DeepCompareIsEqual(const Value& other) const
+    {
+        if (IsString() && other.IsString())
+        {
+            // If we both hold the same ref counted string we don't need to do a full comparison
+            if (AZStd::holds_alternative<AZStd::shared_ptr<const AZStd::string>>(m_value) && m_value == other.m_value)
+            {
+                return true;
+            }
+            return GetString() == other.GetString();
+        }
+
+        if (m_value.index() != other.m_value.index())
+        {
+            return false;
+        }
+
+        return AZStd::visit(
+            [&](auto&& ourValue) -> bool
+            {
+                using Alternative = AZStd::decay_t<decltype(ourValue)>;
+                auto&& theirValue = AZStd::get<AZStd::remove_cvref_t<decltype(ourValue)>>(other.m_value);
```

Comment:
This isn't needed in all branches so the value should only be retrieved when needed.

Reply:
It's only not needed in the null branch, and since it's `if constexpr` it has a good chance of being optimized out since monostate is 0-size (also, notably, doing
```
                if constexpr (AZStd::is_same_v<Alternative, AZStd::monostate>)
                {
                    return true;
                }

                auto&& theirValue = AZStd::get<AZStd::remove_cvref_t<decltype(ourValue)>>(other.m_value);

                if constexpr (AZStd::is_same_v<Alternative, ObjectPtr>)
                // ...
```
leads to a compile error due to "unreachable code" from the first branch) - I'd say it's probably OK for now, but I can always revisit later